### PR TITLE
Refactor the CAS code to move the logic to a handler

### PR DESCRIPTION
**Pull request by clokep**
This refactors the CAS code to move the logic to a separate handler class. It somewhat takes the `SSOAuthHandler` and the logic of each of the servlets and combines them into a `CasHandler` class that is analogous to `SamlHandler`.

It also adds typing information to the various CAS methods.

Based on feedback in https://github.com/matrix-org/synapse/pull/7102#discussion_r397086662

### CODE REVIEW

File: synapse/handlers/cas_handler.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            "service": self._build_service_param(client_redirect_url),
+        }
+        try:
+            body = await self._http_client.get_raw(uri, args)
+        except PartialDownloadError as pde:
+            # Twisted raises this error if the connection is closed,
+            # even if that's being used old-http style to signal end-of-data
+            body = pde.response
+
+        await self._handle_cas_response(request, body, client_redirect_url)
```
Comment (clokep) [inline synapse/handlers/cas_handler.py]: This had a return on it, but `_handle_cas_response` returns an `awaitable` (which is `_on_successful_auth`), which resolves to `None`. It seemed clearer to remove the `return` here?

I believe the servlet should also have the return call removed.

**Comment by clokep** [comment]
This should be ready now, found a couple of other minor things while poking at #7102.

File: synapse/handlers/cas_handler.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        if not registered_user_id:
+            registered_user_id = await self._registration_handler.register_user(
+                localpart=localpart, default_display_name=user_display_name
+            )
+
+        self._auth_handler.complete_sso_login(
+            registered_user_id, request, client_redirect_url
+        )
+
+    def handle_redirect_request(self, client_redirect_url: bytes) -> bytes:
```
Comment (richvdh) [inline synapse/handlers/cas_handler.py]: docstring please? when is this called, what does it return?

Comment (richvdh) [reply synapse/handlers/cas_handler.py]: likewise `handle_ticket_request`

File: synapse/handlers/cas_handler.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+        return ("%s/login?%s" % (self._cas_server_url, args)).encode("ascii")
+
+    async def handle_ticket_request(self, request: SynapseRequest):
+        client_redirect_url = parse_string(request, "redirectUrl", required=True)
+        uri = self._cas_server_url + "/proxyValidate"
+        args = {
+            "ticket": parse_string(request, "ticket", required=True),
+            "service": self._build_service_param(client_redirect_url),
+        }
```
Comment (richvdh) [inline synapse/handlers/cas_handler.py]: generally I'd expect to see this sort of param-parsing in the REST layer. there may be good reasons to put it here though?

File: synapse/handlers/cas_handler.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/cas_handler.py]: Agreed.

I think it's worth noting in the docstring that on a successful request, we'll write a redirect to the `request`, which is why this returns None.

Also we can annotate the function definition to emphasise this:

```
    async def handle_ticket_request(self, request: SynapseRequest) -> None:
```

File: synapse/handlers/cas_handler.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    def _build_service_param(self, client_redirect_url: AnyStr) -> str:
+        return "%s%s?%s" % (
+            self._cas_service_url,
+            "/_matrix/client/r0/login/cas/ticket",
+            urllib.parse.urlencode({"redirectUrl": client_redirect_url}),
+        )
+
+    def _handle_cas_response(
+        self, request: SynapseRequest, cas_response_body: str, client_redirect_url: str
+    ) -> Awaitable:
```
Comment (richvdh) [inline synapse/handlers/cas_handler.py]: I think it's clearer to define this to be `async`, with a `return await` at the end.

(if we're going to annotate it with Awaitable, we should say what type the Awaitable returns: `Awaitable[None]`.)

File: synapse/handlers/cas_handler.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                respond to it with a redirect.
+
+            client_redirect_url: the redirect_url the client gave us when
+                it first started the process.
+
+            user_display_name: if set, and we have to register a new user,
+                we will set their displayname to this.
+
+        Returns:
+            Completes once we have handled the request.
```
Comment (richvdh) [inline synapse/handlers/cas_handler.py]: "Completes" only really makes sense when we are talking about deferreds.

In this case, there's not much to say about the return value; we may as well omit this bit.

File: synapse/handlers/cas_handler.py (hunk 3 above)
Comment (clokep) [reply synapse/handlers/cas_handler.py]: I believe I just followed the pattern of the SAML handler, e.g.:

https://github.com/matrix-org/synapse/blob/229eb81498b0fe1da81e9b5b333a0285acde9446/synapse/handlers/saml_handler.py#L113-L114

Comment (clokep) [reply synapse/handlers/cas_handler.py]: This is easy enough to change and makes sense. 👍

File: synapse/handlers/cas_handler.py (hunk 2 above)
Comment (clokep) [reply synapse/handlers/cas_handler.py]: I added docstrings everywhere. 👍

**Pull request merged**